src/base64url.c

```diff
@@ -1,51 +1,106 @@
+  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
+
+// Maps a character in the Base64url alphabet to its index, per RFC 4648,
+// Table 2. Invalid characters map to 64.
 static const uint8_t ece_base64url_decode_table[] = {
-  255, 255, 255, 255, 255,        255, 255, 255, 255, 255,        255, 255,
-  255, 255, 255, 255, 255,        255, 255, 255, 255, 255,        255, 255,
-  255, 255, 255, 255, 255,        255, 255, 255, 255, 255,        255, 255,
-  255, 255, 255, 255, 255,        255, 255, 255, 255, 62 /* - */, 255, 255,
-  52,  53,  54,  55,  56,         57,  58,  59,  60,  61, /* 0 - 9 */
-  255, 255, 255, 255, 255,        255, 255, 0,   1,   2,          3,   4,
-  5,   6,   7,   8,   9,          10,  11,  12,  13,  14,         15,  16,
-  17,  18,  19,  20,  21,         22,  23,  24,  25, /* A - Z */
-  255, 255, 255, 255, 63 /* _ */, 255, 26,  27,  28,  29,         30,  31,
-  32,  33,  34,  35,  36,         37,  38,  39,  40,  41,         42,  43,
-  44,  45,  46,  47,  48,         49,  50,  51, /* a - z */
-  255, 255, 255, 255,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 52, 53, 54, 55, 56, 57, 58, 59, 60,
+  61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
+  11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64,
+  63, 64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42,
+  43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64,
 };
 
-static inline bool
-ece_base64url_decode_lookup(char c, uint8_t* b) {
-  *b = ece_base64url_decode_table[c & 0x7f];
-  return (*b != 255) && !(*b & ~0x7f);
-}
+// Returns the size of the buffer required to hold the Base64url output,
+// or 0 if `binaryLen` is too large.
+static inline size_t
+ece_base64url_base64_length(size_t binaryLen) {
+  if (binaryLen / 3 > SIZE_MAX / 4) {
+    return 0;
+  }
+  size_t requiredBase64Len = (binaryLen / 3) * 4;
 
-size_t
-ece_base64url_decode(const char* base64, size_t base64Len,
-                     ece_base64url_decode_policy_t paddingPolicy,
-                     uint8_t* decoded, size_t decodedLen) {
-  // Don't decode empty strings.
-  if (!base64Len) {
+  // The final quantum can be 2 or 3 bytes: 1 byte encodes to 2 bytes, and 2
+  // bytes encode to 3 bytes. We also include 1 byte for the NUL terminator.
+  size_t finalLen = 0;
+  switch (binaryLen % 3) {
+  case 1:
+    finalLen = 2;
+    break;
+
+  case 2:
+    finalLen = 3;
+    break;
+  }
+  if (finalLen > SIZE_MAX - requiredBase64Len) {
     return 0;
   }
+  return requiredBase64Len + finalLen;
+}
+
+// Encodes a `binary` quantum into `base64`. A 3-byte quantum encodes to 4
+// bytes, a 2-byte quantum encodes to 3 bytes, and a 1-byte quantum encodes to
+// 2 bytes.
+static inline size_t
+ece_base64url_encode_quantum(const uint8_t* binary, size_t binaryLen,
```

Comment:
Mention that this returns the number of Base64 bytes, for pointer arithmetic. Also, change the return type to `int`.